lib/Language/Souffle/Class.hs

```diff
@@ -0,0 +1,251 @@
+    <> "' to the ProgramFacts type in the Program instance for " <> prog <> ".")
+  CheckContains _ (a ': _) a = ()
+  CheckContains prog (_ ': as) b = CheckContains prog as b
+
+-- | A typeclass for describing a datalog program.
+--
+-- Example usage (assuming the program was generated from path.dl
+-- and contains 2 facts: Edge and Reachable):
+--
+-- @
+-- data Path = Path  -- Handle for the datalog program
+--
+-- instance Program Path where
+--   type ProgramFacts Path = '[Edge, Reachable]
+--   factName = const "path"
+-- @
+class Program a where
+  -- | A type level list of facts that belong to this program.
+  --   This list is used to check that only known facts are added to a program.
+  type ProgramFacts a :: [Type]
+
+  -- | Function for obtaining the name of a Datalog program.
+  --   This has to be the same as the name of the .dl file (minus the extension).
+  --
+  -- It uses a 'Proxy' to select the correct instance.
+  programName :: Proxy a -> String
+
+-- | A typeclass for data types representing a fact in datalog.
+class Marshal.Marshal a => Fact a where
+  -- | Function for obtaining the name of a fact
+  --   (has to be the same as described in the Datalog program).
+  --
+  -- It uses a 'Proxy' to select the correct instance.
+  --
+  -- Example usage:
+  --
+  -- @
+  -- instance Fact Edge where
+  --   factName = const "edge"
+  -- @
+  factName :: Proxy a -> String
+
+---- | Helper typeclass for collecting facts into a container-like structure.
+----   The order of returned facts is unspecified for performance reasons.
+----   Only used internally.
+--class CollectFacts m c where
+--  type CollectFactsIterator m c :: Type
+--  collectFacts :: Marshal.Marshal a
+--               => Int
+--               -> CollectFactsIterator m c
+--               -> IO (c a)
+
+-- | A mtl-style typeclass for Souffle-related actions.
+class Monad m => MonadSouffle m where
+  type Handler m :: Type -> Type
+  {- | Initializes a Souffle program.
+
+     The action will return 'Nothing' if it failed to load the Souffle program.
+     Otherwise it will return a 'Handle' that can be used in other functions
+     in this module.
+  -}
+  init :: Program prog => prog -> m (Maybe (Handler m prog))
+
+  -- | Runs the Souffle program.
+  run :: Handler m prog -> m ()
+
+  -- | Sets the number of CPU cores this Souffle program should use.
+  setNumThreads :: Handler m prog -> Word64 -> m ()
+
+  -- | Gets the number of CPU cores this Souffle program should use.
+  getNumThreads :: Handler m prog -> m Word64
+
+  -- | Load all facts from files in a certain directory.
+  loadFiles :: Handler m prog -> FilePath -> m ()
+
+  -- | Write out all facts of the program to CSV files
+  --   (as defined in the Souffle program).
+  writeFiles :: Handler m prog -> m ()
+
+  -- | Returns all facts of a program. This function makes use of type inference
+  --   to select the type of fact to return.
+  getFacts :: (Fact a, ContainsFact prog a)
```

Comment:
I removed the CollectFacts typeclass. Do you know a good way to bring it back?

Reply:
Let me know, when you start to work on this. I also want to try to bring them back. My original approach was to try to use a mult parameter type class, but got mysterious type errors. If I have time at the weekend I'll also try to look into this, but only one of us should be working on this.

Reply:
Idk what the issue is here yet exactly, but the previous version did rely heavily on type inference, so if it didn't know to return either a list or a vector, it might have given you a type error there.
Let me check into this later today; then we can figure out a way to add the functionality back again.

Reply:
Well, I managed to get old behavior back. With a slightly more complex signature:
https://github.com/luc-tielen/souffle-haskell/pull/7/commits/e00edcdd8053e2f537c99964c43a75658815b6b4#diff-a1698a52bd67230f916c05d598f21fe0R82

Basically an associated type that maps to a typeclass constraint. And then you write 1 for both Interpreted and Compiled modes, with 2 instances each for vectors and lists. 💥